### PR TITLE
Feature/giftcard random number alpha numeric

### PR DESCRIPTION
**Pull request by joshua1234511**
Generate random gift-card number
So it is not easy to generate by customers
This is helpful if printing cards and selling or gifts to customers

### CODE REVIEW

**Comment by joshua1234511** [comment]
https://www.screencast.com/t/ILokfElvxee

**Comment by daN4cat** [comment]
It looks a good idea but could you make it optional? Not everybody wants it random as some could have preprinted and serialised cards.

File: database/giftcard_random.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+-- alter giftcard field number to be varchar
+
+ALTER TABLE `ospos_giftcards` CHANGE `giftcard_number` `giftcard_number` VARCHAR(255) NULL;
```
Comment (daN4cat) [inline database/giftcard_random.sql]: please add this at the end of database/3.0.2_3.1.0.sql script and also add the correct field type to tables.sql in ospos_giftcards table (I'll regen the database script correctly before merging once done)

File: application/controllers/Giftcards.php [hunk 2]
```diff
@@ -75,9 +75,14 @@ public function view($giftcard_id = -1)
 	
 	public function save($giftcard_id = -1)
 	{
+		if($giftcard_id == -1):
```
Comment (daN4cat) [inline application/controllers/Giftcards.php]: please use the standard way in this code:

if(..)
{
}
else
{
}

we prefer to maintain consistency across the code

File: application/models/Giftcard.php [hunk 3]
```diff
@@ -256,5 +256,32 @@ public function update_giftcard_value($giftcard_number, $value)
...
+
+		return ($this->db->get()->num_rows() == 1);
+	}
+
+	/*
+	Generate unique gift card name/number
+	*/
+	public function generate_unique_giftcard_name($value)
+	{
+		$value = str_replace('.', 'DE', $value);
```
Comment (daN4cat) [inline application/models/Giftcard.php]: I don't get, what is this supposed to do?

File: application/controllers/Giftcards.php [hunk 4]
```diff
@@ -75,9 +75,14 @@ public function view($giftcard_id = -1)
 	
 	public function save($giftcard_id = -1)
 	{
+		if($giftcard_id == -1):
+			$giftcard_number = $this->Giftcard->generate_unique_giftcard_name($this->input->post('value'));
+		else:
+			$giftcard_number = $this->input->post('giftcard_number');
```
Comment (daN4cat) [inline application/controllers/Giftcards.php]: what happens if somebody changes code to an existing giftcard and clashes with an existing one?

File: application/views/giftcards/form.php [hunk 5]
```diff
@@ -17,8 +17,9 @@
 			</div>
 		</div>
 
+	<?php if($giftcard_number != ''): ?>
```
Comment (daN4cat) [inline application/views/giftcards/form.php]: does this mean that a new giftcard will not allow inputting a number?
It's a limitation compared to the past behaviour.

File: application/controllers/Giftcards.php (hunk 4 above)
Comment (daN4cat) [inline application/controllers/Giftcards.php]: please disregard this comment, I checked better your video and I could see the scenario implemented properly.

File: application/models/Giftcard.php [hunk 6]
```diff
@@ -256,5 +256,32 @@ public function update_giftcard_value($giftcard_number, $value)
...
+		$random = bin2hex(openssl_random_pseudo_bytes(3));
+		$giftcard_name = (string)$random.'-'.$value;
+		if($this->exists_gitcard_name($giftcard_name)){
```
Comment (daN4cat) [inline application/models/Giftcard.php]: we use the "old style"

if(..)
{
}

File: application/controllers/Sales.php [hunk 7]
```diff
@@ -195,10 +195,12 @@ public function set_email_receipt()
 	public function add_payment()
 	{
 		$data = array();
-		$this->form_validation->set_rules('amount_tendered', 'lang:sales_amount_tendered', 'trim|required|callback_numeric');
 
 		$payment_type = $this->input->post('payment_type');
-
+		if($payment_type != $this->lang->line('sales_giftcard'))
```
Comment (daN4cat) [inline application/controllers/Sales.php]: please add 
{
}

File: database/giftcard_random.sql (hunk 1 above)
Comment (joshua1234511) [inline database/giftcard_random.sql]: Did this

File: application/controllers/Giftcards.php (hunk 2 above)
Comment (joshua1234511) [inline application/controllers/Giftcards.php]: did this

File: application/controllers/Giftcards.php (hunk 4 above)
Comment (joshua1234511) [inline application/controllers/Giftcards.php]: ignored as working

File: application/models/Giftcard.php (hunk 3 above)
Comment (joshua1234511) [inline application/models/Giftcard.php]: Here im added the value at end : so easy to read    but if decimal value:     the . dot replacing : 

With what to replace 
ill use - instead

File: application/views/giftcards/form.php (hunk 5 above)
Comment (joshua1234511) [inline application/views/giftcards/form.php]: Integrated old method also 
User can add input also

File: application/models/Giftcard.php (hunk 6 above)
Comment (joshua1234511) [inline application/models/Giftcard.php]: followed

File: application/controllers/Sales.php (hunk 7 above)
Comment (joshua1234511) [inline application/controllers/Sales.php]: did

**Comment by joshua1234511** [comment]
@daN4cat 
Need help for this function : 
public function get_max_number()
	{
		$this->db->select_max('giftcard_number');

		return $this->db->get('giftcards')->row();
	}

I need to make it to this: 
so that it work for both : SELECT MAX(`giftcard_number`) FROM `ospos_giftcards` WHERE `giftcard_number` REGEXP '^[0-9]+$'


The above query works fine but how to convert to this format

File: application/models/Giftcard.php (hunk 3 above)
Comment (joshua1234511) [inline application/models/Giftcard.php]: watch this eg: 
https://www.screencast.com/t/SrGGe7hnZX4

Comment (daN4cat) [inline application/models/Giftcard.php]: ok got it, thanks

**Comment by daN4cat** [comment]
ok, good you fixed. I was checking it.

**Comment by joshua1234511** [comment]
thanks for checking: 👍 
Ill upload the video of working now

**Comment by daN4cat** [comment]
Checking the code and looks good. So I can merge now.

**Comment by joshua1234511** [comment]
Doing some final testing of cases will put here once done testing

**Comment by joshua1234511** [comment]
Did random test 
Working all good 
https://www.screencast.com/t/qc3PxLxN 
It can be merged:

**Comment by daN4cat** [comment]
Merged and thank you for the contribution.

**Pull request closed without merging**